### PR TITLE
feature: Linear Regression online spmd support

### PR DESCRIPTION
**Pull request by DDJHB**
# Description
Adding spmd support for online Linear Regression

### CODE REVIEW

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_dpc.cpp [hunk 1]
```diff
@@ -14,129 +14,32 @@
...
+                                         const detail::descriptor_base<Task>& desc,
+                                         const detail::train_parameters<Task>& params,
+                                         const partial_train_result<Task>& input) {
+    return finalize_train_kernel_norm_eq_impl<Float, Task>(ctx)(desc, params, input);
 }
 
 template <typename Float, typename Task>
 struct finalize_train_kernel_gpu<Float, method::norm_eq, Task> {
-    train_result<Task> operator()(const context_gpu& ctx,
+    train_result<Task> operator()(const be::context_gpu& ctx,
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_dpc.cpp:38]: perhaps it makes sense to rename it to bk

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_impl_dpc.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        result.set_coefficients(coefficients);
+    }
+
+    return result;
+}
+
+template class finalize_train_kernel_norm_eq_impl<float, task::regression>;
+template class finalize_train_kernel_norm_eq_impl<double, task::regression>;
+
+} // namespace oneapi::dal::linear_regression::backend
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_impl_dpc.cpp:112]: add newline

File: cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_impl_dpc.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    const auto feature_count = ext_feature_count - compute_intercept;
+
+    const pr::ndshape<2> xtx_shape{ ext_feature_count, ext_feature_count };
+
+    const auto xtx_nd =
+        pr::table2ndarray<Float>(q, input.get_partial_xtx(), sycl::usm::alloc::device);
+    const auto xty_nd = pr::table2ndarray<Float, pr::ndorder::f>(q,
+                                                                 input.get_partial_xty(),
+                                                                 sycl::usm::alloc::device);
+
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_impl_dpc.cpp:56]: I guess you have to do:
    {
        ONEDAL_PROFILER_TASK(allreduce_xtx, q);
        comm_.allreduce(xtx_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
    }
        {
        ONEDAL_PROFILER_TASK(allreduce_xty, q);
        comm_.allreduce(xty_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
    }

File: cpp/oneapi/dal/algo/linear_regression/test/online_spmd.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    this->generate(777);
+
+    this->set_rank_count(GENERATE(1, 2, 4));
+    std::int64_t n_blocks = GENERATE(1, 3, 10);
+
+    this->run_and_check_linear_online_spmd(this->get_rank_count(), n_blocks);
+}
+
+} // namespace oneapi::dal::linear_regression::test
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/linear_regression/test/online_spmd.cpp:126]: add newline

File: cpp/oneapi/dal/algo/linear_regression/test/online_spmd.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    void set_rank_count(std::int64_t rank_count) {
+        n_rank = rank_count;
+    }
+
+    std::int64_t get_rank_count() {
+        return n_rank;
+    }
+
+    void generate_dimensions() {
+        this->t_count_ = GENERATE(307, 12999);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/linear_regression/test/online_spmd.cpp:46]: it makes sense to little bit reduce the number of tests for online linreg, check that it takes locally not too much time

Comment (DDJHB) [reply cpp/oneapi/dal/algo/linear_regression/test/online_spmd.cpp:46]: this is already reduced set of tests, takes 3.9 seconds. do you think this is good enough or should i reduce further?

Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/algo/linear_regression/test/online_spmd.cpp:46]: Probably increase up to ~20seconds

**Review by Alexandr-Solovev — APPROVED** [review]
LGTM! Nice work! Well done!

**Comment by DDJHB** [comment]
latest internal ci run: https://ecmd.jf.intel.com/commander/link/jobDetails/jobs/ef44f3df-0baa-f11d-9fc5-a4bf010d0e2e
somehow "/intelci:run" does not trigger/attach it to the "checks"

**Pull request merged**